File: HW/HW12/main.java
package HW.HW12;

import java.util.ArrayList;

public class main {

    public static void main(String[] args) {
        main solution = new main();

        int n = 5;
        int[] wells3 = {5, 8, 2, 1, 9};
        int[][] pipes3 = {{1, 2, 3}, {2, 3, 2}, {3, 4, 1}, {4, 5, 5}, {5, 1, 2}};
        System.out.println(solution.minCostToSupplyWater(n, wells3, pipes3));
    }

    public static ArrayList<Integer> findMinPipe(int[][] pipes, ArrayList<Integer> done) {
        int lowpipe = Integer.MAX_VALUE;
        ArrayList<Integer>  result = new ArrayList<Integer>();
        int lowHousePipe = -10;
    
        for (int[] edge : pipes) {
            if (done.contains(edge[0]) && !done.contains(edge[1]) && edge[2] < lowpipe) {
                lowpipe = edge[2];
                lowHousePipe = edge[1];
            } else if (done.contains(edge[1]) && !done.contains(edge[0]) && edge[2] < lowpipe) {
                lowpipe = edge[2];
                lowHousePipe = edge[0];
            }
        }
    
        result.add(lowpipe);
        result.add(lowHousePipe);

        return result;
    }
    
    public static ArrayList<Integer> findMinWell(int[] wells, ArrayList<Integer> done){
        int lowwell = Integer.MAX_VALUE;
        ArrayList<Integer>  result = new ArrayList<Integer>();
        int bestHouseWell = -10;
        for (int i = 0; i < wells.length; i++) {

            if (!done.contains(i + 1) && wells[i] < lowwell) {
                lowwell = wells[i];
                bestHouseWell = i + 1;
            }
        }

        result.add(lowwell);
        result.add(bestHouseWell);
        
        return result;

    }
    
    public int minCostToSupplyWater(int n, int[] wells, int[][] pipes) {
        int totalCost = 0;
        ArrayList<Integer> done = new ArrayList<>();
        
        while (n > done.size()) {
            int lowwell = findMinWell(wells, done).get(0);
            int bestHouseWell = findMinWell(wells, done).get(1);

            int lowpipe = findMinPipe(pipes, done).get(0);
            int lowHousePipe = findMinPipe(pipes, done).get(1);

            if (lowwell < lowpipe) {
                done.add(bestHouseWell);

                totalCost += lowwell;

            } else if(lowpipe < lowwell){
                done.add(lowHousePipe);
                totalCost += lowpipe;

            }
            else if(lowpipe == lowwell){
                done.add(lowHousePipe);
                totalCost += lowpipe;
            }
        }
        
        return totalCost;
    }
}

File: HW/HW12/main.py
from typing import List, Set

class Solution:

    def minCostToSupplyWater(self, n: int, wells: List[int], pipes: List[List[int]]) -> int:

        total_cost = 0

        done = set()

        while len(done) < n:

            lowwell, besthousewell = self.find_min_well(wells, done)

            lowpipe, lowhousepipe = self.find_min_pipe(pipes, done)


            if lowwell < lowpipe:
                done.add(besthousewell)
                total_cost += lowwell

            elif lowpipe < lowwell:

                done.add(lowhousepipe)
                total_cost += lowpipe
            else:

                done.add(lowhousepipe)

                total_cost += lowpipe

        return total_cost
    

    def find_min_well(self, wells: List[int], done: Set[int]) -> int:

        lowwell = float('inf')
        besthousewell = -10


        for i, well in enumerate(wells):
            if i + 1 not in done and well < lowwell:

                lowwell = well

                besthousewell = i + 1
        return lowwell, besthousewell

    def find_min_pipe(self, pipes: List[List[int]], done: Set[int]) -> int:
        lowpipe = float('inf')

        lowhousepipe = -10


        for pipe in pipes:
            if pipe[0] in done and pipe[1] not in done and pipe[2] < lowpipe:
                lowpipe = pipe[2]

            
                lowhousepipe = pipe[1]

            elif pipe[1] in done and pipe[0] not in done and pipe[2] < lowpipe:

                lowpipe = pipe[2]

                lowhousepipe = pipe[0]

        return lowpipe, lowhousepipe

n = 5
wells = [5, 8, 2, 1, 9]

pipes = [[1, 2, 3], [2, 3, 2], [3, 4, 1], [4, 5, 5], [5, 1, 2]]


solution = Solution()
print(solution.minCostToSupplyWater(n, wells, pipes))
